source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -150,6 +161,37 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
   return {action, bytes_read, end_stream};
 }
 
+void SslSocket::onPrivateKeyMethodComplete() {
+  ASSERT(isThreadSafe());
+  ASSERT(async_handshake_in_progress_);
+  async_handshake_in_progress_ = false;
+
+  ENVOY_CONN_LOG(debug, "async handshake complete", callbacks_->connection());
+  if (!handshake_complete_) {
+    // It's possible that the async call comes in later, but the handshake has been retried from
+    // doWrite or similar. In that case the error handling has been done synchronously. */
+    PostIoAction result = doHandshake();
+    if (async_handshake_in_progress_ || result != PostIoAction::KeepOpen) {
```

Comment:
The code was trying to prepare for two possible issues in the providers.

The first case was someone returning `ssl_private_key_retry` from the completion function even **after** being called from `onPrivateKeyMethodComplete()` callback. In this case the `async_handshake_in_progress_` flag would be become `true` again. The idea was that we could detect the error even though the resulting action would be `PostIoAction::KeepOpen`. However, as you say, we can just agree that providers are not allowed to return `ssl_private_key_retry` if they have already called the callback. Anyway in my opinion we shouldn't make the callback function idempotent if the operation is still pending, but request that it must be called exactly once.

The second case was to guard about the race condition in the provider where the key operation is ready but `doHandshake()` is called from `doWrite()` instead of `onPrivateKeyMethodComplete()` callback. This is why the `if (handshake_complete_)` path was there. However, since we are directly calling the handshake function again from the callback, we can just say that it's an error if the provider returns `ssl_private_key_success` or `ssl_private_key_failure` from the completion function **before** it has called the callback.

If we go with these (very reasonable) semantics then we can very well have the simpler form of the callback which you suggest. The example providers already work this way. I'll do the fix.